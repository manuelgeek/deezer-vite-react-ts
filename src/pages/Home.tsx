import InfiniteScroll from 'react-infinite-scroll-component'
import GridEmptyState from '../components/GridEmptyState'
import HomeHeader from '../components/HomeHeader'
import LoadingState from '../components/LoadingState'
import TrackCard from '../components/TrackCard'
import { useSearchTrack } from '../hooks/useSearchTrack'

function Home() {
  const { search, data, isLoading, debouncedSearchTerm, fetchNextData } =
    useSearchTrack()

  return (
    <div className="w-full min-h-screen">
      <HomeHeader />
      <div className="flex justify-center">
        <div className="w-full">
          {debouncedSearchTerm && (
            <h3 className="text-left py-4 px-4">
              Search results for :{' '}
              <span className="font-bold">{search ? `"${search}"` : ''}</span>
            </h3>
          )}
          <div className="flex flex-wrap">
            {isLoading ? (
              <LoadingState />
            ) : data && debouncedSearchTerm ? (
              <InfiniteScroll
                dataLength={data?.data?.length}
                next={fetchNextData}
                hasMore={Boolean(data?.next)}
                loader={<LoadingState />}
                endMessage={
                  <div className="flex w-full justify-center text-xl">
                    <b>Yay! You have seen it all</b>
                  </div>
                }
              >
                <div className="flex flex-wrap">
                  {data?.data?.map((track) => (
                    <div key={track.id} className="w-full p-4 md:w-1/5">
                      <TrackCard track={track} />
                    </div>
                  ))}
                </div>
              </InfiniteScroll>
            ) : (
              <GridEmptyState
                message={
                  debouncedSearchTerm
                    ? `No results for ${debouncedSearchTerm}. We can't find any item matching your search`
                    : 'Type in the search bar to search your favorite artists...'
                }
              />
            )}
          </div>
        </div>
      </div>
    </div>
  )
}

export default Home
