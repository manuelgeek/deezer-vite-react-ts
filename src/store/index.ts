import { configureStore } from '@reduxjs/toolkit'
import searchReducer from './searchSlice'

export const store = configureStore({
  reducer: {
    search: searchReducer,
  },
})

// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>
// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
export type AppDispatch = typeof store.dispatch
