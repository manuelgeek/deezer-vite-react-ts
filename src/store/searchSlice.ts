/* eslint-disable no-param-reassign */
import { createSlice } from '@reduxjs/toolkit'

const initialState: { value: string } = {
  value: '',
}

export const searchSlice = createSlice({
  name: 'search',
  initialState,
  reducers: {
    clearSearch: (state) => {
      state.value = ''
    },
    addSearchTerm: (state, { payload }: { payload: string }) => {
      state.value = payload
    },
  },
})

export const { clearSearch, addSearchTerm } = searchSlice.actions

export default searchSlice.reducer
