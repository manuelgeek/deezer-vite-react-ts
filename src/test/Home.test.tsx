/* eslint-disable no-undef */
import React from 'react'
import renderer, { act } from 'react-test-renderer'
import { render, screen, waitFor } from '@testing-library/react'
import Home from '../pages/Home'
import { store } from '../store'
import { addSearchTerm } from '../store/searchSlice'
import { ReduxProvider } from './utils/test-utils'
import tracks from './mockData/tracks.json'
// eslint-disable-next-line import/no-unresolved
import '@testing-library/jest-dom/extend-expect'

// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
global.fetch = jest.fn(() =>
  Promise.resolve({
    json: () => Promise.resolve(tracks),
  })
)

afterEach(() => {
  store.dispatch(addSearchTerm(''))
})

describe('Jest Snapshot Home', () => {
  it('Matches DOM Snapshot on empty', () => {
    const domTree = renderer
      .create(
        <ReduxProvider reduxStore={store}>
          <Home />
        </ReduxProvider>
      )
      .toJSON()
    expect(domTree).toMatchSnapshot()
  })

  it('Matches DOM Snapshot on search', async () => {
    const domTree = renderer.create(
      <ReduxProvider reduxStore={store}>
        <Home />
      </ReduxProvider>
    )

    act(() => {
      store.dispatch(addSearchTerm('test'))
    })

    await domTree.update(
      <ReduxProvider reduxStore={store}>
        <Home />
      </ReduxProvider>
    )

    expect(domTree.toJSON()).toMatchSnapshot()
  })

  it('Test search artist input empty', async () => {
    render(
      <ReduxProvider reduxStore={store}>
        <Home />
      </ReduxProvider>
    )

    expect(
      screen.getByText(
        'Type in the search bar to search your favorite artists...'
      )
    ).toBeInTheDocument()
  })

  it('Test search artist with "test" in input', async () => {
    store.dispatch(addSearchTerm('test'))

    render(
      <ReduxProvider reduxStore={store}>
        <Home />
      </ReduxProvider>
    )

    await waitFor(() =>
      expect(screen.getByText('Search results for :')).toBeInTheDocument()
    )
  })
})
