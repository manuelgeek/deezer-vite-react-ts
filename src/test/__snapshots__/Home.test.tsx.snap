// Vitest Snapshot v1

exports[`Jest Snapshot Home > Matches DOM Snapshot on empty 1`] = `
<div
  className="w-full min-h-screen"
>
  <header
    className="flex justify-center bg-gray-200 border-b border-gray-300 py-2 md:py-3"
  >
    <div
      className="w-full items-center px-4 md:px-12"
    >
      <div
        className="flex items-center justify-between border-l-2 px-4"
      >
        <svg
          className="h-4 w-4"
          fill="none"
          stroke="currentColor"
          strokeWidth={2}
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M21 21l-6-6m2-5a7 7 0 11-14 0 7 7 0 0114 0z"
            strokeLinecap="round"
            strokeLinejoin="round"
          />
        </svg>
        <input
          className="outline-none py-4 px-4 bg-transparent w-3/4"
          data-testid="search"
          id="search"
          onChange={[Function]}
          placeholder="Type here to search ...."
          type="text"
          value=""
        />
        <button
          className="text-gray-400 bg-transparent hover:bg-gray-200 hover:text-gray-900 rounded-lg text-sm p-1.5 ml-auto inline-flex items-center dark:hover:bg-gray-600 dark:hover:text-white"
          onClick={[Function]}
          type="button"
        >
          <svg
            aria-hidden="true"
            className="w-5 h-5"
            fill="currentColor"
            viewBox="0 0 20 20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clipRule="evenodd"
              d="M4.293 4.293a1 1 0 011.414 0L10 8.586l4.293-4.293a1 1 0 111.414 1.414L11.414 10l4.293 4.293a1 1 0 01-1.414 1.414L10 11.414l-4.293 4.293a1 1 0 01-1.414-1.414L8.586 10 4.293 5.707a1 1 0 010-1.414z"
              fillRule="evenodd"
            />
          </svg>
          <span
            className="sr-only"
          >
            Clear text
          </span>
        </button>
      </div>
    </div>
  </header>
  <div
    className="flex justify-center"
  >
    <div
      className="w-full"
    >
      <div
        className="flex flex-wrap"
      >
        <div
          className="w-full py-24 md:py-40 items-center justify-center"
          role="status"
        >
          <svg
            className="inline mr-2 w-24 h-24 text-gray-200 animate-spin dark:text-gray-600 fill-blue-600"
            fill="none"
            viewBox="0 0 100 101"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M100 50.5908C100 78.2051 77.6142 100.591 50 100.591C22.3858 100.591 0 78.2051 0 50.5908C0 22.9766 22.3858 0.59082 50 0.59082C77.6142 0.59082 100 22.9766 100 50.5908ZM9.08144 50.5908C9.08144 73.1895 27.4013 91.5094 50 91.5094C72.5987 91.5094 90.9186 73.1895 90.9186 50.5908C90.9186 27.9921 72.5987 9.67226 50 9.67226C27.4013 9.67226 9.08144 27.9921 9.08144 50.5908Z"
              fill="currentColor"
            />
            <path
              d="M93.9676 39.0409C96.393 38.4038 97.8624 35.9116 97.0079 33.5539C95.2932 28.8227 92.871 24.3692 89.8167 20.348C85.8452 15.1192 80.8826 10.7238 75.2124 7.41289C69.5422 4.10194 63.2754 1.94025 56.7698 1.05124C51.7666 0.367541 46.6976 0.446843 41.7345 1.27873C39.2613 1.69328 37.813 4.19778 38.4501 6.62326C39.0873 9.04874 41.5694 10.4717 44.0505 10.1071C47.8511 9.54855 51.7191 9.52689 55.5402 10.0491C60.8642 10.7766 65.9928 12.5457 70.6331 15.2552C75.2735 17.9648 79.3347 21.5619 82.5849 25.841C84.9175 28.9121 86.7997 32.2913 88.1811 35.8758C89.083 38.2158 91.5421 39.6781 93.9676 39.0409Z"
              fill="currentFill"
            />
          </svg>
          <span
            className="sr-only"
          >
            Loading...
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Jest Snapshot Home > Matches DOM Snapshot on search 1`] = `
<div
  className="w-full min-h-screen"
>
  <header
    className="flex justify-center bg-gray-200 border-b border-gray-300 py-2 md:py-3"
  >
    <div
      className="w-full items-center px-4 md:px-12"
    >
      <div
        className="flex items-center justify-between border-l-2 px-4"
      >
        <svg
          className="h-4 w-4"
          fill="none"
          stroke="currentColor"
          strokeWidth={2}
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M21 21l-6-6m2-5a7 7 0 11-14 0 7 7 0 0114 0z"
            strokeLinecap="round"
            strokeLinejoin="round"
          />
        </svg>
        <input
          className="outline-none py-4 px-4 bg-transparent w-3/4"
          data-testid="search"
          id="search"
          onChange={[Function]}
          placeholder="Type here to search ...."
          type="text"
          value="test"
        />
        <button
          className="text-gray-400 bg-transparent hover:bg-gray-200 hover:text-gray-900 rounded-lg text-sm p-1.5 ml-auto inline-flex items-center dark:hover:bg-gray-600 dark:hover:text-white"
          onClick={[Function]}
          type="button"
        >
          <svg
            aria-hidden="true"
            className="w-5 h-5"
            fill="currentColor"
            viewBox="0 0 20 20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clipRule="evenodd"
              d="M4.293 4.293a1 1 0 011.414 0L10 8.586l4.293-4.293a1 1 0 111.414 1.414L11.414 10l4.293 4.293a1 1 0 01-1.414 1.414L10 11.414l-4.293 4.293a1 1 0 01-1.414-1.414L8.586 10 4.293 5.707a1 1 0 010-1.414z"
              fillRule="evenodd"
            />
          </svg>
          <span
            className="sr-only"
          >
            Clear text
          </span>
        </button>
      </div>
    </div>
  </header>
  <div
    className="flex justify-center"
  >
    <div
      className="w-full"
    >
      <div
        className="flex flex-wrap"
      >
        <div
          className="w-full py-24 md:py-40 items-center justify-center"
          role="status"
        >
          <svg
            className="inline mr-2 w-24 h-24 text-gray-200 animate-spin dark:text-gray-600 fill-blue-600"
            fill="none"
            viewBox="0 0 100 101"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M100 50.5908C100 78.2051 77.6142 100.591 50 100.591C22.3858 100.591 0 78.2051 0 50.5908C0 22.9766 22.3858 0.59082 50 0.59082C77.6142 0.59082 100 22.9766 100 50.5908ZM9.08144 50.5908C9.08144 73.1895 27.4013 91.5094 50 91.5094C72.5987 91.5094 90.9186 73.1895 90.9186 50.5908C90.9186 27.9921 72.5987 9.67226 50 9.67226C27.4013 9.67226 9.08144 27.9921 9.08144 50.5908Z"
              fill="currentColor"
            />
            <path
              d="M93.9676 39.0409C96.393 38.4038 97.8624 35.9116 97.0079 33.5539C95.2932 28.8227 92.871 24.3692 89.8167 20.348C85.8452 15.1192 80.8826 10.7238 75.2124 7.41289C69.5422 4.10194 63.2754 1.94025 56.7698 1.05124C51.7666 0.367541 46.6976 0.446843 41.7345 1.27873C39.2613 1.69328 37.813 4.19778 38.4501 6.62326C39.0873 9.04874 41.5694 10.4717 44.0505 10.1071C47.8511 9.54855 51.7191 9.52689 55.5402 10.0491C60.8642 10.7766 65.9928 12.5457 70.6331 15.2552C75.2735 17.9648 79.3347 21.5619 82.5849 25.841C84.9175 28.9121 86.7997 32.2913 88.1811 35.8758C89.083 38.2158 91.5421 39.6781 93.9676 39.0409Z"
              fill="currentFill"
            />
          </svg>
          <span
            className="sr-only"
          >
            Loading...
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
